server/coinlock/coinlocker.go

```diff
@@ -0,0 +1,222 @@
+	lockedCoins        map[coinIDKey]struct{}
+	lockedCoinsByOrder map[order.OrderID][]CoinID
+}
+
+// NewAssetCoinLocker constructs a new AssetCoinLocker.
+func NewAssetCoinLocker() *AssetCoinLocker {
+	return &AssetCoinLocker{
+		lockedCoins:        make(map[coinIDKey]struct{}),
+		lockedCoinsByOrder: make(map[order.OrderID][]CoinID),
+	}
+}
+
+// CoinLocked indicates if a coin identifier (e.g. UTXO) is locked.
+func (ac *AssetCoinLocker) CoinLocked(coin CoinID) bool {
+	ac.coinMtx.RLock()
+	_, locked := ac.lockedCoins[coinIDKey(coin)]
+	ac.coinMtx.RUnlock()
+	return locked
+}
+
+// OrderCoinsLocked lists the coin IDs (e.g. UTXOs) locked by an order.
+func (ac *AssetCoinLocker) OrderCoinsLocked(oid order.OrderID) []CoinID {
+	ac.coinMtx.RLock()
+	defer ac.coinMtx.RUnlock()
+	return ac.lockedCoinsByOrder[oid]
+}
+
+// UnlockOrderCoins unlocks any coins backing order.
+func (ac *AssetCoinLocker) UnlockOrderCoins(oid order.OrderID) {
+	ac.coinMtx.Lock()
+	coins := ac.lockedCoinsByOrder[oid]
+	for i := range coins {
+		delete(ac.lockedCoins, coinIDKey(coins[i]))
+	}
+	ac.coinMtx.Unlock()
+}
+
+// LockCoins locks all coins (e.g. UTXOS) connected with certain orders.
+func (ac *AssetCoinLocker) LockCoins(orderCoins map[order.OrderID][]CoinID) {
+	ac.coinMtx.Lock()
+	for oid, coins := range orderCoins {
+		ac.lockedCoinsByOrder[oid] = coins
+		for i := range coins {
+			// _, locked := ac.lockedCoins[coins[i]]
+			// if locked {
+			// 	panic(fmt.Sprintf("coin already locked: %v", coins[i]))
+			// }
+			ac.lockedCoins[coinIDKey(coins[i])] = struct{}{}
+		}
+	}
+	ac.coinMtx.Unlock()
+}
+
+// LockOrdersCoins locks all coins associated with certain orders.
+func (ac *AssetCoinLocker) LockOrdersCoins(orders []order.Order) {
+	ac.coinMtx.Lock()
+	for _, ord := range orders {
+		coinIDs := ord.CoinIDs()
+		if len(coinIDs) == 0 {
+			continue // e.g. CancelOrder
+		}
+
+		ac.lockedCoinsByOrder[ord.ID()] = coinIDs
+		for i := range coinIDs {
+			// _, locked := ac.lockedCoins[coinIDs[i]]
+			// if locked {
+			// 	panic(fmt.Sprintf("coin already locked: %v", coinIDs[i]))
+			// }
+			ac.lockedCoins[coinIDKey(coinIDs[i])] = struct{}{}
+		}
+	}
+
+	ac.coinMtx.Unlock()
+}
+
+type DEXCoinLocker struct {
+	masterLocks map[uint32]*MasterCoinLocker
+}
+
+func NewDEXCoinLocker(assets []uint32) *DEXCoinLocker {
+	masterLocks := make(map[uint32]*MasterCoinLocker, len(assets))
+	for _, asset := range assets {
+		masterLocks[asset] = NewMasterCoinLocker()
+	}
+
+	return &DEXCoinLocker{masterLocks}
+}
+
+func (c *DEXCoinLocker) CoinLocked(asset uint32, coin string) bool {
+	locker := c.masterLocks[asset]
+	if locker == nil {
+		panic(fmt.Sprintf("unknown asset %d", asset))
+	}
+
+	return locker.CoinLocked(CoinID(coin))
+}
+
+func (c *DEXCoinLocker) OrderCoinsLocked(asset uint32, oid order.OrderID) []CoinID {
+	locker := c.masterLocks[asset]
+	if locker == nil {
+		panic(fmt.Sprintf("unknown asset %d", asset))
+	}
+
+	return locker.OrderCoinsLocked(oid)
+}
```

Comment:
These may or may not be used by the highest level DEX component, hence not commented, but kept here to facilitate the next steps.